### ISSUE TITLE
SameDiff calculates unnecessary gradients

### ISSUE THREAD

**Issue by orausch** [deeplearning4j/deeplearning4j#8333]
#### Issue Description
I want to calculate gradients for only specific variables in my samediff graph.
In the following code, `z.gradient().getArr()` should be null, since I specified to only backpropagate the gradient for `x`.
https://gist.github.com/orausch/361c3582fa899697d7e407b8bc913a31

**Actual Output:**
```
w.gradient(): 8.0000
x.gradient(): 8.0000
z.gradient(): 4.0000
```
**Expected Output:**
```
w.gradient(): 8.0000
x.gradient(): 8.0000
z.gradient(): null
```
#### Version Information
`1.0.0-beta5`

If this is not the right way to do what I want to do, please let me know how to do it.

**Comment by AlexDBlack** [comment]
This is expected behaviour.
```
SDVariable loss = sd.loss().meanSquaredError("loss", sd.constant(Nd4j.zeros(1)), w.mul(z));
```

Note you have `w.mul(z)` - clearly, `w` gradient is required, so this multiplication operation is required to be differentiated. Let O=w.mul(z), then the mul_bp operation has inputs (w, z, dL/dO), and outputs (dL/dw and dL/dz).

In general SameDiff automatically determines the minimal subgraph necessary to calculate the required gradients; however this minimal subgraph is at the operation level, not the variable level. Consequently we get gradients for all (floating point) inputs for operations that need to be differentiated.
